### PR TITLE
Replace unknown remote function

### PR DESCRIPTION
**Pull request by sheldak**
I've prepared that PR to propose extending code actions by replacing unknown remote functions. For example, the following code:

```elixir
defmodule Example do
  def foo do
    var = Enum.counts([2, 3])
  end
end
```

contains an unknown function: `counts/1`. After this change, we would get 2 possible Quick Fixes:

```
Replace unknown function with 'Enum.concat'
Replace unknown function with 'Enum.count'
```

that rewrite our code to:

```elixir
defmodule Example do
  def foo do
    var = Enum.concat([2, 3])
  end
end
```

or 

```elixir
defmodule Example do
  def foo do
    var = Enum.count([2, 3])
  end
end
```

respectively.

In our example, Quick Fixes were generated from the following Elixir warning:
```
Enum.counts/1 is undefined or private. Did you mean:

      * concat/1
      * concat/2
      * count/1
      * count/2
```

I find such a feature very convenient, so I'm waiting for comments and feedback.

### CODE REVIEW

**Comment by lukaszsamson** [comment]
Nice work @sheldak . The problem with that approach is that it's error prone. We recently had to revert a similar text based code transformation (https://github.com/elixir-lsp/elixir-ls/pull/775). A better option would be to express the code changes in terms of AST transformations. See https://github.com/elixir-lsp/elixir-ls/pull/773

**Comment by sheldak** [comment]
@lukaszsamson thanks for the update! So I will wait with contributing until #773 is merged.

**Comment by scohen** [comment]
@sheldak #773 has landed.

**Comment by sheldak** [comment]
@scohen That's great! I will definitely come back to this feature soon

**Comment by sheldak** [comment]
@scohen I've updated that so now it should fit the new way of doing code actions

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.CodeAction, as: CodeActionResult
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Diagnostic
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.TextEdit
+  alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Workspace
+  alias ElixirLS.LanguageServer.Experimental.SourceFile
+
+  @pattern ~r/(.*)\/(.*) is undefined or private. .*:\n(.*)/s
+
+  @spec pattern() :: Regex.t()
+  def pattern, do: @pattern
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:17]: this exposes private implementation details,  I have an example of how to handle this without doing so below.

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  end
+
+  defp separate_module_from_name(full_name) do
+    {name, module} =
+      full_name
+      |> String.split(".")
+      |> Enum.map(&String.to_atom/1)
+      |> List.pop_at(-1)
+
+    {module, name}
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:51]: nit, module isn't actually a module, it's a list, and it will be missing the leading `Elixir` that all elixir modules have.

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+          end)
+
+        {:ok, suggestions}
+
+      _ ->
+        :error
+    end
+  end
+
+  defp extract_line(%Diagnostic{} = diagnostic) do
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:75]: nit:  `extract_start_line`

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  @pattern ~r/(.*)\/(.*) is undefined or private. .*:\n(.*)/s
+
+  @spec pattern() :: Regex.t()
+  def pattern, do: @pattern
+
+  @spec apply(SourceFile.t(), Diagnostic.t()) :: [CodeActionResult.t()]
+  def apply(source_file, diagnostic) do
+    with {:ok, module, name} <- extract_function(diagnostic.message),
+         {:ok, suggestions} <- extract_suggestions(diagnostic.message),
+         one_based_line = extract_line(diagnostic),
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:23]: nit: I'm not a big fan of putting things in a with clause that can't fail, you don't have any guarantees about what `one_based_line` can be here, it can be nil, for example.

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+  defp construct_title(module_list, function_name) do
+    module_string =
+      module_list
+      |> Enum.map(fn module ->
+        module
+        |> Atom.to_string()
+        |> String.trim_leading("Elixir.")
+      end)
+      |> Enum.join(".")
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:143]: i _think_ you can just do 
```elixir
module_string = 
Enum.map(module_list, fn module -> 
  module |> Module.split() |> Enum.join(".")
end)
```

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    "Replace function with #{module_string}.#{function_name}"
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:145]: this is going to be pretty long in editors. I'd suggest dropping "function" and make it "Replace with"

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+              )
+            end)
+
+          {:ok, replies}
+      end
+    end
+  end
+
+  defp text_edits_per_suggestion(line_text, line_ast, module, name, suggestions) do
+    Enum.reduce(suggestions, {:ok, []}, fn
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:110]: this should probably be `reduce_while` rather than just `reduce` , since you're stopping after the first error. 
Taken with the suggestion to stop returning `ok` tuples above, this would be:
```elixir
result = 
Enum.reduce_while(suggestions, [], fn suggestion, acc -> 
  case CodeMod.ReplaceRemoteFunction.text_edits(...) do 
    {:ok, []} -> {:cont, text_edits_per_suggesions}
    {:ok, edits} {:cont, [{edits, suggestion} | edits_per_suggestion}]}
    _ -> {:halt, :error}
 end
end)
case result do 
  :error -> :error 
  edits -> {:ok, edits}
end
```

This is a little more clear and stops execution when an error is encountered.

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+              _ -> []
+            end
+          end)
+
+        {:ok, suggestions}
+
+      _ ->
+        :error
+    end
+  end
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:73]: can this actually fail in a meaningful way? Returning `ok` tuples really complicates some of the other functions here. Would it be better to just to return an empty list if there are no suggestions?

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_with_underscore.ex [hunk 9]
```diff
@@ -5,28 +5,26 @@ defmodule ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceWithUn
...
   alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Diagnostic
   alias ElixirLS.LanguageServer.Experimental.Protocol.Types.TextEdit
   alias ElixirLS.LanguageServer.Experimental.Protocol.Types.Workspace
   alias ElixirLS.LanguageServer.Experimental.SourceFile
 
-  @spec apply(CodeAction.t()) :: [CodeActionResult.t()]
-  def apply(%CodeAction{} = code_action) do
-    source_file = code_action.source_file
-    diagnostics = get_in(code_action, [:context, :diagnostics]) || []
+  @pattern ~r/variable "([^"]+)" is unused/
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_with_underscore.ex:14]: nit: Pattern is a much less descriptive name than `variable_re`, but we should not be exposing this anyways.

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs [hunk 10]
```diff
@@ -71,12 +70,24 @@ defmodule ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceWithUn
...
   test "produces no actions if the name or variable is not found" do
     assert {_, action} = code_action("other_var = 6", "/project/file.ex", 1, "not_found")
-    assert [] = apply(action)
+
+    source_file = action.source_file
+    diagnostics = get_in(action, [:context, :diagnostics])
+
+    Enum.each(diagnostics, fn diagnostic ->
+      assert [] = ReplaceWithUnderscore.apply(source_file, diagnostic)
+    end)
```
Comment (scohen) [inline apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:79]: why this change? 
Placing asserts in a loop makes them very hard to debug and reason about.

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs [hunk 11]
```diff
@@ -95,31 +111,12 @@ defmodule ElixirLS.LanguageServer.Experimental.Provider.CodeAction.ReplaceWithUn
...
-  test "handles empty diagnostics" do
-    assert {_, action} = code_action("other_var = 6", "/project/file.ex", 1, "not_found")
-
-    action = put_in(action, [:context, :diagnostics], [])
-
-    assert [] = apply(action)
+    Enum.each(diagnostics, fn diagnostic ->
+      assert [] = ReplaceWithUnderscore.apply(source_file, diagnostic)
+    end)
   end
```
Comment (scohen) [inline apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:120]: why did you delete these tests? They were added for a reason (editors sometimes do this)

**Review by scohen — CHANGES_REQUESTED** [review]

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            end)
+
+          replies =
+            Enum.map(edits_per_suggestion, fn {text_edits, function_name} ->
+              CodeActionResult.new(
+                title: construct_title(module, function_name),
+                kind: :quick_fix,
+                edit: Workspace.Edit.new(changes: %{source_file.uri => text_edits})
+              )
+            end)
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:102]: it looks like these can be pipelined or done in one go

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+        [_ | _] ->
+          edits_per_suggestion =
+            Enum.map(edits_per_suggestion, fn {text_edits, suggestion} ->
+              text_edits = Enum.map(text_edits, &update_line(&1, one_based_line))
+              {text_edits, suggestion}
+            end)
+
+          replies =
+            Enum.map(edits_per_suggestion, fn {text_edits, function_name} ->
```
Comment (scohen) [inline apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:96]: the variable is called `edits_per_suggestion` but in the inner function, it's called `function_name`

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex (hunk 4 above)
Comment (sheldak) [reply apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:23]: Actually, I'm matching against the `Types.Diagnostic` module in the `extract_line` function and it seems like none of the keys in `diagnostic.range.start.line` have an optional type, so why isn't it guaranteed that `one_based_line` is an integer?

Also, I think changing `extract_line` to

```elixir
defp extract_line(%Diagnostic{} = diagnostic) do
  {:ok, diagnostic.range.start.line}
end
```
does not give us anything. If it is not guaranteed that `diagnostic.range.start.line` can't be nil, we would just get `{:ok, nil}` in `apply` function and still have the same problem.

By `one_based_line = extract_line(diagnostic),` I'm avoiding nested `with`s.

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex (hunk 5 above)
Comment (sheldak) [reply apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:143]: I wrongly assumed that this function gets a list of modules (e.g. [ElixirLS, LanguageServer, Experimental]), but instead it gets a list of atoms (e.g. [:ElixirLS, :LanguageServer, :Experimental]) so `String.trim_leading("Elixir.")` seems to be unnecessary. The following code should work:

```elixir
 module_string =
    module_list
    |> Enum.map(&Atom.to_string/1)
    |> Enum.join(".")
```

`Module.split` does not work with atoms

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex (hunk 8 above)
Comment (sheldak) [reply apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:73]: That's right, an empty list would be better here

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex (hunk 13 above)
Comment (sheldak) [reply apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:96]: I will rename `function_name` with `suggestion`

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 10 above)
Comment (sheldak) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:79]: I still wanted to test only `ReplaceWithUnderscore.apply` (instead of `CodeAction.handle`) here and that required having both `source_file` and `diagnostics`. I'm looping to call `apply` for every diagnostic

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 11 above)
Comment (sheldak) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:120]: I think I should have created a new test module `Provider.CodeActionTest` and put them there. They just were not applicable for `ReplaceWithUnderscore.apply` since we don't pass the whole action anymore (I decided to extract the file and diagnostics before calling `apply` since we need them in both code actions)

File: apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex (hunk 5 above)
Comment (scohen) [reply apps/language_server/lib/language_server/experimental/provider/code_action/replace_remote_function.ex:143]: then this is just
```elixir
module_string = Enum.map_join(module_list, ".", &Atom.to_string/1)
```

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 10 above)
Comment (scohen) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:79]: I'm confused, this didn't call `CodeAction.handle`

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 11 above)
Comment (scohen) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:120]: > I think I should have created a new test module Provider.Handlers.CodeActionTest and put them there.

Agreed. 

> I decided to extract the file and diagnostics 

I think that change has made things a lot more difficult. why not just let the code actions get the request and return a list of code actions that they fine (or an empty list if they don't find anything?

This is what I'm talking about: https://github.com/elixir-lsp/elixir-ls/pull/776/files#r1131330271

**Review by scohen — CHANGES_REQUESTED** [review]

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 10 above)
Comment (sheldak) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs:79]: I will change it back anyway because of https://github.com/elixir-lsp/elixir-ls/pull/776#discussion_r1138573992

File: apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs (hunk 11 above)
Comment (scohen) [reply apps/language_server/test/experimental/provider/code_action/replace_with_underscore_test.exs]: > there is no point in creating Provider.Handlers.CodeActionTest, isn't it?

That's the idea. Code that's so simple, it can't be wrong.

File: apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+    assert edits
+           |> Enum.zip(expected_edits)
+           |> Enum.all?(fn {%TextEdit{new_text: new_text}, %TextEdit{new_text: expected_new_text}} ->
+             new_text == expected_new_text
+           end)
+
+    assert Enum.all?(edits, fn edit -> edit.range.start.line == line end)
+    assert Enum.all?(edits, fn edit -> edit.range.end.line == line end)
+  end
```
Comment (scohen) [inline apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs:82]: These kind of asserts are really difficult to reason about. It's hard to see intrinsically if they're correct or not. 
I'm a big fan of behavior based testing, and I think it makes more sense to see that applying the code action to a segment of code achieves the desired result. 

The good news is that `SourceFile` _almost_ has what you need to accomplish this. All that's required is the following function head to be added (provide you alias TextEdit):  
```elixir
  defp apply_change(%__MODULE__{} = source, %TextEdit{} = change) do
    apply_change(source, change.range, change.new_text)
  end
```

and you'll be able to rewrite this in terms of a document transformation rather than comparing text edits, so you could do:
```elixir
expected_doc = """
Enum.count(a)
"""
assert expected_doc == actual_doc
|> code_action("/project/file.ex", 0)
|> apply_selected_action(0)
```
`apply_selected_action` would select the given action from the ones generated, select that index and send the list of `%TextEdit{}`s to `SourceFile.apply_changes`

which is easier to reason about and understand, and should make this test a bit more concise and digestable. 
Also, I'd pull it out later, and make a CodeAction.Case that everyone developing CodeActions could use.

**Review by scohen — COMMENTED** [review]
One more comment regarding the tests, and we're good to merge

File: apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+             |> code_action("/project/file.ex", 3)
+             |> apply_selected_action(0)
+
+    expected_doc = ~S[
+      defmodule MyModule do
+        def my_func(a) do
+          Enum.concat(a)
+        end
+      end
+    ] |> Document.new()
```
Comment (scohen) [inline apps/language_server/test/experimental/provider/code_action/replace_remote_function_test.exs:209]: wow, this works very well; I think the structural == will even work with documents. Very nice.

**Review by scohen — APPROVED** [review]
